Upgrade to Lucene 9.9.0

ChrisHegarty opened the pull request:
Upgrade to Lucene 9.9.0

TODO:
 - [x] Update TransportVersion constant ([75c5b87](https://github.com/elastic/elasticsearch/pull/102782/commits/75c5b870b4eb68fe0049a2e5f5395c938d031c11))
 - [x] Review and Update IndexVersion constant ([12906cd](https://github.com/elastic/elasticsearch/pull/102782/commits/12906cd8f33dae7c8d1169b3aa6d55f3f4bfe065))
 - [x] Remove lucene RC maven repo
 - [x] Verify Changelogs (there is currently just one!) ([36869d4](https://github.com/elastic/elasticsearch/pull/102782/commits/36869d417d54e8e19ccf5ec613468059e55cd958))

Note: This PR will merge it's commits into _main_ (rather than squashing), in order to preserve history and already-reviewed changes.

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-b3e67403aaf

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

benwtrent committed:
Adds new max_inner_product vector similarity function (#99527)

Adds new max_inner_product vector similarity function. This differs from dot_product in the following ways:

Doesn't require vectors to be normalized
Scales the similarity between vectors differently to prevent negative scores

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-f01ff9d1f51

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-be57460b060

jpountz committed:
Fix compilation after refactoring of TermStates.

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

jpountz committed:
Refactor changes to IndexVersion. (#99312)

This adds a version for the Lucene upgrade and adjusts some tests.

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-d3a3391d225

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-0fb47cd44a6

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-0fb47cd44a6

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-350de210c36

iverase committed:
Add RandomAccessInput#length to SeekTrackingDirectoryWrapper (#99962)

Fix compiling error after adding a new method to RandomAccessInput
interface.

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-350de210c36

javanna committed:
Merge branch 'main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-9ba7f2dc4bc

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-7c1d1147beb

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-7c1d1147beb

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-bab19260197

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-8c994d1e7c4

javanna committed:
Merge branch 'main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-cccaa7e7298

javanna committed:
Merge branch 'main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-b85aeb3a4fa

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-b85aeb3a4fa

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-b85aeb3a4fa

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-455d4152d31

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-823af4931aa

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-aa968f96d6c

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-4533dcea4ec

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-a5f94b1e81e

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-a1bb48aa426

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-28255de5bee

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

romseygeek committed:
Fix compilation (#100903)

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-ba26abcaee9

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-18bb826564b

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-40fbff02f1e

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-6fc3483e4fa

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-07a76555d9e

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-3292aca1f45

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-ad0f00a6cb2

javanna committed:
Merge branch 'main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-1cb1a14cc84

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-170f594daea

javanna committed:
Update references to Lucene95Codec

This is needed following https://github.com/apache/lucene/pull/12685 and https://github.com/apache/lucene/pull/12582

javanna committed:
Fix spotless issues

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

javanna committed:
Address compile error in BWCCodec

javanna committed:
Fix compile errors on Lucene62SegmentInfoFormat and Lucene50SegmentInfoFormat

These are needed after https://github.com/apache/lucene/pull/12685

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-2bb54320c33

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

javanna committed:
Fix arguments order in BWCCodec#wrap

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-063cfa7a85c

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-5fe48424a25

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-5b26498ec72

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

javanna committed:
Switch ContextIndexSearcher to use Lucene's TaskExecutor (#101537)

We have contributed back to Lucene the changes we had made around running
concurrent tasks. These include waiting for all tasks to finish when an
exception is thrown, as well as not starting tasks when one of the
previously run tasks throws an exception.

The execution of concurrent tasks is now generalized within Lucene and
exposed through a TaskExecutor that can be retrieved from the
IndexSearcher and used to run tasks. We still have customizations that
require us to override some of the search method, but with this change
we rely on standard Lucene's behaviour for running concurrent tasks.

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-44479b3b48b

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-83727a88e62

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-ab9cbe5aa00

javanna committed:
Merge branch 'main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-71b3e4c97fb

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-6684da1908a

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-f7c1de55999

javanna committed:
Merge branch 'main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-d9109907bca

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

javanna committed:
Fix compile errors (#101874)

IndexDiskUsageAnalyzer and IndexDiskUsageAnalyzerTests, as well as CompletionFieldMapper, CompletionFieldMapperTests and CompletionStatsCacheTests need adjusting after apache/lucene#12741 , to refer to the latest postings format.
KuromojiTokenizerFactory needs adjusting after apache/lucene#12390

benwtrent committed:
Adjust SortField comparators to use new Pruning API (#101983)

Introduced in https://github.com/apache/lucene/pull/12405

We should account for the changes in our overrides and API. Now, to indicate that no skipping can occur, we utilize `Pruning.NONE`.

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-9a0245333ff

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-448e6112954

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-448e6112954

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

javanna committed:
Resolve compile error in DenseVectorFieldMapper (#102066)

* Resolve compile error in DenseVectorFieldMapper

A change in Lucene99HnswVectorsFormat requires that we adapt our code, see
https://github.com/apache/lucene/pull/12729

* Add new Lucene file extensions

* Fixing format name check

---------

Co-authored-by: Benjamin Trent <4357155+benwtrent@users.noreply.github.com>

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-0eda40a371b

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-910c721e065

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

benwtrent committed:
Fix failing tests due to new query to string format (#102186)

There have been some changes around range query toString format. This commit adjusts tests expecting particular outputs from the previous Lucene version

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-b13e4a121ab

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-b13e4a121ab

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-2e8dfac07e2

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-85e4deab437

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-85e4deab437

javanna committed:
Merge branch 'main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-6cd78318eab

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

saikatsarkar056 committed:
Gather vector_operation count for knn search (#102032)

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-175031da6ae

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-1138a4064e2

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-ea8b6476ed3

javanna committed:
Merge branch 'main' into lucene_snapshot

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

javanna committed:
Introduce transport version for the next lucene upgrade (#102587)

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-02677650e19

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-c367ee3ea1a

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-c367ee3ea1a

javanna committed:
Merge branch 'main' into lucene_snapshot

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

javanna committed:
spotless

benwtrent committed:
Fix IndexDiskUsageAnalyzerTests with vectors (#102320)

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-41da5c0b6a9

Commit:
Merge remote-tracking branch 'origin/main' into lucene_snapshot

Commit:
[Automated] Update Lucene snapshot to 9.9.0-snapshot-a6d788e1138

ChrisHegarty committed:
Merge 'origin/main' into lucene_snapshot

benwtrent committed:
Add byte quantization for float vectors in HNSW (#102093)

Adds new `quantization_options` to `dense_vector`. This allows for
vectors to be automatically quantized to `byte` when indexed.

Example:

```
PUT vectors
{
  "mappings": {
    "properties": {
      "my_vector": {
        "type": "dense_vector",
        "index": true,
        "index_options": {
          "type": "int8_hnsw"
        }
      }
    }
  }
}
```

When querying, the query vector is automatically quantized and used when
querying the HNSW graph. This reduces the memory required to only `25%`
of what was previously required for `float` vectors at a slight loss of
accuracy.

This is currently only available when `index: true` and when using
`hnsw`

ChrisHegarty committed:
Use lucene 9.0.0 RC1

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

ChrisHegarty committed:
Fix gradle verification metadata

elasticsearchmachine commented:
Hi @ChrisHegarty, I've created a changelog YAML for you.

ChrisHegarty committed:
Update docs/changelog/102782.yaml

ChrisHegarty committed:
Use a real UPGRADE_TO_LUCENE_9_9 TransportVersion id

ChrisHegarty committed:
Fix IndexVersion to use Lucene 9.9

idegtiarenko commented on server/src/main/java/org/elasticsearch/index/IndexVersions.java:
```diff
@@ -89,7 +89,7 @@ private static IndexVersion def(int id, Version luceneVersion) {
     public static final IndexVersion SPARSE_VECTOR_IN_FIELD_NAMES_SUPPORT = def(8_500_002, Version.LUCENE_9_7_0);
     public static final IndexVersion UPGRADE_LUCENE_9_8 = def(8_500_003, Version.LUCENE_9_8_0);
-    public static final IndexVersion ES_VERSION_8_12 = def(8_500_004, Version.LUCENE_9_8_0);
+    public static final IndexVersion UPGRADE_LUCENE_9_9 = def(8_500_004, Version.LUCENE_9_9_0);
```
No objection from my side regarding this renaming

ChrisHegarty replied:
Thanks @idegtiarenko. Rename and version bump :-)

ChrisHegarty committed:
Merge branch 'main' into lucene_snapshot_9_9

benwtrent committed:
Fixing byte quantized search test flakiness

Quantized search result scores can be slightly different depending on
if they are merged into a single segment, etc.

ChrisHegarty committed:
Bump to lucene 9.9.0 RC2

ChrisHegarty committed:
Merge 'main' into lucene_snapshot_9_9

benwtrent commented:
@javanna you are correct, let me commit it really quick

benwtrent committed:
Adding changelog for PR #102093

ChrisHegarty committed:
Merge 'main' into lucene_snapshot_9_9

ChrisHegarty commented:
@elasticsearchmachine run elasticsearch-ci/part-1

benwtrent committed:
Fix search.vectors/41_knn_search_byte_quantized/Knn search with mip flakiness

ChrisHegarty committed:
Merge branch 'main' into lucene_snapshot_9_9

ChrisHegarty committed:
Update to 9.9.0 RC

ChrisHegarty committed:
Merge 'main' into lucene_snapshot_9_9

ChrisHegarty committed:
Merge 'main' into lucene_snapshot_9_9

ChrisHegarty committed:
Merge 'main' into lucene_snapshot_9_9
